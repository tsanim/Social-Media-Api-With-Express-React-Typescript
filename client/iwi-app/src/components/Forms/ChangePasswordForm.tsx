import React, { useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faLock, faKey } from '@fortawesome/free-solid-svg-icons';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import getFieldStyles from '../../utils/getFieldStyles';
import ChangePasswordFormProps from '../../interfaces/Components/Forms/ChangePasswordFormProps.interface';
import { ChangePassword } from '../../interfaces/Components/EditUserInformation/EditUserInfoProps.interface';

const validationSchema = Yup.object().shape({
    oldPassword: Yup.string()
        .required('Required'),
    newPassword: Yup.string()
        .matches(/^.*(?=.{6,})(?=.*[a-zA-Z])[a-zA-Z0-9]+$/, 'Password must contain at least one letter and at least one digit!')
        .min(8, 'Password must be at least 8 digits')
        .required('Required'),
});

function ChangePasswordForm({ changePasswordHandler, createSnackbar, errors: serverErrors, resetErrorsHandler }: ChangePasswordFormProps) {
    const handleSubmit = (form: ChangePassword) => {
        changePasswordHandler(form);

        createSnackbar({
            message: 'Password is changed!',
            timeout: 3000,
        });

        //clear all errors 
        resetErrorsHandler();
    }

    //clear all errors after component unmount, so the same errors not showing up on another component
    useEffect(() => {
        return () => {
            resetErrorsHandler();
        }
    }, [resetErrorsHandler]);

    return (
        <Formik
            onSubmit={handleSubmit}
            initialValues={{
                oldPassword: '',
                newPassword: ''
            }}
            validationSchema={validationSchema}
        >
            {
                ({ errors }: any) => {
                    return (
                        <Form className="editForm">
                            {
                                (serverErrors.length > 0 && (<p className="error">{serverErrors[serverErrors.length - 1].message}</p>))
                            }
                            <span>
                                <FontAwesomeIcon icon={faKey} />
                                <Field
                                    type="password"
                                    name="oldPassword"
                                    placeholder="Old password..."
                                    style={getFieldStyles(errors, 'oldPassword')}
                                />
                                <ErrorMessage className="error-message" name="oldPassword" component="div" />
                            </span>
                            <span>
                                <FontAwesomeIcon icon={faLock} />
                                <Field
                                    type="password"
                                    name="newPassword"
                                    placeholder="New password..."
                                    style={getFieldStyles(errors, 'newPassword')}
                                />
                                <ErrorMessage className="error-message" name="newPassword" component="div" />
                            </span>
                            <input type="submit" value="CHANGE" />
                        </Form>
                    )
                }
            }
        </Formik>

    )
}

ChangePasswordForm.propTypes = {
    changePasswordHandler: PropTypes.func,
    createSnackbar: PropTypes.func,
    errors: PropTypes.array,
    resetErrorsHandler: PropTypes.func
}

export default wrapComponent(ChangePasswordForm);