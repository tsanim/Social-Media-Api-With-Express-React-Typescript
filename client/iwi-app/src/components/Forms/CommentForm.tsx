import React from 'react';
import useForms from '../../hooks/useForms';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import CommentFormProps from '../../interfaces/Components/Forms/CommentFormProps.interface';

export function CommentForm({ postId, makeCommentHandler, createSnackbar }: CommentFormProps) {
    const { handleSubmit, handleChangeInput, inputs } = useForms(() => {
        //check for empty text and send message to user
        if (inputs.text === '' || !inputs.text) {
            createSnackbar({
                message: 'You can not make comment without text!',
                timeout: 3000,
            });
        } else {
            makeCommentHandler({ ...inputs, postId })
        }
    });

    return (
        <form className="commentForm" onSubmit={handleSubmit}>
            <textarea
                placeholder="Write a comment..."
                name="text"
                id="comment"
                value={inputs.text || ''}
                onChange={handleChangeInput}
            />
            <input disabled={inputs.text === '' || !inputs.text} type="submit" value="Comment" />
        </form>
    )
}

CommentForm.propTypes = {
    postId: PropTypes.string,
    makeCommentHandler: PropTypes.func,
    createSnackbar: PropTypes.func
}

export default wrapComponent(CommentForm);
