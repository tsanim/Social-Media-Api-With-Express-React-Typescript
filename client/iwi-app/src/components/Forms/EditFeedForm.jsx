import React from 'react';
import useForms from '../../hooks/useForms';
import PropTypes from 'prop-types';

function EditFeedForm({ feedId, text, editUserPostHandler, handleShowEditForm }) {
    const { handleSubmit, handleChangeInput, inputs } = useForms((e) => {
        editUserPostHandler(inputs, feedId);
        handleShowEditForm();
    });

    return (
        <form id="editFeedForm" onSubmit={handleSubmit}>
            <textarea
                placeholder="Write something..."
                name="text"
                id="feed"
                value={inputs.text || text}
                onChange={handleChangeInput}
            />
            <input type="submit" value="SAVE" />
            <button onClick={handleShowEditForm}>Undo</button>
        </form>
    )
}

EditFeedForm.propTypes = {
    feedId: PropTypes.string,
    text: PropTypes.string,
    editUserPostHandler: PropTypes.func,
    handleShowEditForm: PropTypes.func,
}

export default EditFeedForm;