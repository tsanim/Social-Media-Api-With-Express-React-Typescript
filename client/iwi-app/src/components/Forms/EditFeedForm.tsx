import React from 'react';
import useForms from '../../hooks/useForms';
import PropTypes from 'prop-types';
import EditFeedFormProps from '../../interfaces/Components/Forms/EditFeedFormProps.interface';

function EditFeedForm({ feedId, text, editUserPostHandler, handleShowEditForm }: EditFeedFormProps) {
    const { handleSubmit, handleChangeInput, inputs } = useForms(() => {
        editUserPostHandler(inputs, feedId);
        handleShowEditForm();
    });

    return (
        <form id="editFeedForm" onSubmit={handleSubmit}>
            <textarea
                placeholder="Write something..."
                name="text"
                id="feed"
                value={inputs.text || text}
                onChange={handleChangeInput}
            />
            <input type="submit" value="SAVE" />
            <button onClick={handleShowEditForm}>Undo</button>
        </form>
    )
}

EditFeedForm.propTypes = {
    feedId: PropTypes.string,
    text: PropTypes.string,
    editUserPostHandler: PropTypes.func,
    handleShowEditForm: PropTypes.func,
}

export default EditFeedForm;