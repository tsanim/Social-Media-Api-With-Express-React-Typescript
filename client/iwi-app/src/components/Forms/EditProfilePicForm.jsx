import React, { Component } from 'react';
import URI from '../../config/config';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';

const fd = new FormData();

export class EditProfilePictureForm extends Component {
    state = {
        imagePreviewUrl: '',
    }

    handleChangeInput = (e) => {
        const files = e.target.files;

        if (files && files.length > 0) {
            let reader = new FileReader();
            let file = e.target.files[0];

            fd.append(e.target.name, file, e.target.name);

            reader.onload = () => {
                this.setState({
                    imagePreviewUrl: reader.result
                });
            }

            reader.readAsDataURL(file);
        }
    }

    handleSubmit = (e) => {
        e.preventDefault();

        this.props.changeProfilePicHandler(fd);

        this.props.createSnackbar({
            message: 'Profile picture is changed!',
            timeout: 3000,
        });
    }

    render() {
        return (
            <form className="editForm" encType="multipart/form-data" onSubmit={this.handleSubmit}>
                <figure>
                    <img className="profileImg" src={this.state.imagePreviewUrl || `${URI}/feed/image/${this.props.imageId}`} alt="userPic" />
                    <div className="uploadImage">
                        <label htmlFor="file"><img className="icon" src="https://cdn.iconscout.com/icon/free/png-256/gallery-44-267592.png" alt="uploadImg" />Change
                      Profile Picture</label>
                        <input type="file" id="file" name="avatar" onChange={this.handleChangeInput} />
                    </div>
                </figure>
                <input type="submit" value="SAVE" />
            </form>
        )
    }
}

EditProfilePictureForm.propTypes = {
    imageId: PropTypes.string,
    changeProfilePicHandler: PropTypes.func,
    createSnackbar: PropTypes.func,
}

export default wrapComponent(EditProfilePictureForm);