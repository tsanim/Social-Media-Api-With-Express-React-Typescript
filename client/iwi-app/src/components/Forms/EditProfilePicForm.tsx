import React, { Component } from 'react';
import URI from '../../config/config';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import EditProfilePictureFormProps from '../../interfaces/Components/Forms/EditProfilePicFormProps.interface';

const fd = new FormData();

export class EditProfilePictureForm extends Component<EditProfilePictureFormProps, { imagePreviewUrl: string }> {
    state = {
        imagePreviewUrl: '',
    }

    handleChangeInput = (e: React.ChangeEvent<HTMLTextAreaElement> | React.ChangeEvent<HTMLInputElement>) => {
        const files = (e.target as HTMLInputElement).files;

        if (files && files.length > 0) {
            let reader = new FileReader();
            let file = (e.target as HTMLInputElement).files[0];

            fd.append(e.target.name, file, e.target.name);

            reader.onload = () => {
                this.setState({
                    imagePreviewUrl: reader.result as string
                });
            }

            reader.readAsDataURL(file);
        }
    }

    handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        this.props.changeProfilePicHandler(fd);

        this.props.createSnackbar({
            message: 'Profile picture is changed!',
            timeout: 3000,
        });
    }

    render() {
        return (
            <form className="editForm" encType="multipart/form-data" onSubmit={this.handleSubmit}>
                <figure>
                    <img className="profileImg" src={this.state.imagePreviewUrl || `${URI}/feed/image/${this.props.imageId}`} alt="userPic" />
                    <div className="uploadImage">
                        <label htmlFor="file"><img className="icon" src="https://cdn.iconscout.com/icon/free/png-256/gallery-44-267592.png" alt="uploadImg" />Change
                      Profile Picture</label>
                        <input type="file" id="file" name="avatar" onChange={this.handleChangeInput} />
                    </div>
                </figure>
                <input type="submit" value="SAVE" />
            </form>
        )
    }
}

// EditProfilePictureForm.propTypes = {
//     imageId: PropTypes.string,
//     changeProfilePicHandler: PropTypes.func,
//     createSnackbar: PropTypes.func,
// }

export default wrapComponent(EditProfilePictureForm);