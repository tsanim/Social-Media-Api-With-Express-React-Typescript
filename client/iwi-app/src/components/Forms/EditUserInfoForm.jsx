import React from 'react';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import getFieldStyles from '../../utils/getFieldStyles';

const initEditInfoState = {
    form: {
        username: '',
        firstName: '',
        lastName: '',
    }
}

const validationSchema = Yup.object().shape({
    username: Yup.string()
        .min(3, 'Username must be at least 3 symbols!'),
    firstName: Yup.string()
        .min(2, 'First name must be at least 2 symbols!'),
    lastName: Yup.string()
        .min(2, 'Last name must be at least 2 symbols!'),
});

export function FormikEditForm({ errors }) {
    return (
        <Form className="editForm">
            <span>
                <Field
                    type="text"
                    name="username"
                    placeholder="Username..."
                    id="username"
                    style={getFieldStyles(errors, 'username')}
                />
                <ErrorMessage className="error-message" name="username" component="div" />
            </span>
            <span>
                <Field
                    type="text"
                    name="firstName"
                    placeholder="First Name..."
                    id="fistName"
                    style={getFieldStyles(errors, 'firstName')}
                />
                <ErrorMessage className="error-message" name="firstName" component="div" />
            </span>
            <span>
                <Field
                    type="text"
                    name="lastName"
                    placeholder="Last Name..."
                    id="lastName"
                    style={getFieldStyles(errors, 'lastName')}
                />
                <ErrorMessage className="error-message" name="lastName" component="div" />
            </span>
            <input type="submit" value="EDIT" />
        </Form>
    )
}

export function EditUserInfoForm({ editUserInfoHandler, createSnackbar }) {
    const handleSubmit = (form) => {
        
        editUserInfoHandler(form);

        createSnackbar({
            message: 'User info succesfully updated!',
            timeout: 3000,
        });
    }

    return (
        <Formik
            onSubmit={handleSubmit}
            initialValues={initEditInfoState}
            validationSchema={validationSchema}

        >
            { FormikEditForm }
        </Formik>
    )
}

EditUserInfoForm.propTypes = {
    editUserInfoHandler: PropTypes.func,
    createSnackbar: PropTypes.func,
}

export default wrapComponent(EditUserInfoForm);