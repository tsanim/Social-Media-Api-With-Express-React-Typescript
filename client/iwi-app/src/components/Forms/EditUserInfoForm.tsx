import React from 'react';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import getFieldStyles from '../../utils/getFieldStyles';
import EditUserInfoFormProps from '../../interfaces/Components/Forms/EditUserInfoFormProps.interface';
import { EditUserInformation } from '../../interfaces/Components/EditUserInformation/EditUserInfoProps.interface';

const initEditInfoState = {
    form: {
        username: '',
        firstName: '',
        lastName: '',
    }
}

const validationSchema = Yup.object().shape({
    username: Yup.string()
        .min(3, 'Username must be at least 3 symbols!'),
    firstName: Yup.string()
        .min(2, 'First name must be at least 2 symbols!'),
    lastName: Yup.string()
        .min(2, 'Last name must be at least 2 symbols!'),
});

export function FormikEditForm({ errors }: any) {
    return (
        <Form className="editForm">
            <span>
                <Field
                    type="text"
                    name="username"
                    placeholder="Username..."
                    id="username"
                    style={getFieldStyles(errors, 'username')}
                />
                <ErrorMessage className="error-message" name="username" component="div" />
            </span>
            <span>
                <Field
                    type="text"
                    name="firstName"
                    placeholder="First Name..."
                    id="fistName"
                    style={getFieldStyles(errors, 'firstName')}
                />
                <ErrorMessage className="error-message" name="firstName" component="div" />
            </span>
            <span>
                <Field
                    type="text"
                    name="lastName"
                    placeholder="Last Name..."
                    id="lastName"
                    style={getFieldStyles(errors, 'lastName')}
                />
                <ErrorMessage className="error-message" name="lastName" component="div" />
            </span>
            <input type="submit" value="EDIT" />
        </Form>
    )
}

export function EditUserInfoForm({ editUserInfoHandler, createSnackbar }: EditUserInfoFormProps) {
    const handleSubmit = (form: EditUserInformation) => {
        
        editUserInfoHandler(form);

        createSnackbar({
            message: 'User info succesfully updated!',
            timeout: 3000,
        });
    }

    return (
        <Formik
            onSubmit={handleSubmit}
            initialValues={initEditInfoState}
            validationSchema={validationSchema}

        >
            { FormikEditForm }
        </Formik>
    )
}

EditUserInfoForm.propTypes = {
    editUserInfoHandler: PropTypes.func,
    createSnackbar: PropTypes.func,
}

export default wrapComponent(EditUserInfoForm);