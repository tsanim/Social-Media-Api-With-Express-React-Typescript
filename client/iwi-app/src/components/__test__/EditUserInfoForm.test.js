import React from 'react';
import { shallow } from 'enzyme';
import { Formik, Form, Field, ErrorMessage } from 'formik';

import { EditUserInfoForm, FormikEditForm } from '../Forms/EditUserInfoForm';

function editUserInfoFormComponentSetup() {
    const props = {
        editUserInfoHandler: jest.fn(),
        createSnackbar: jest.fn(),
    }

    const enzymeWrapper = shallow(<EditUserInfoForm {...props} />)

    return {
        props,
        enzymeWrapper
    }
}

describe('EditUserInfoForm Component', () => {
    it('should render initial layout', () => {
        const { enzymeWrapper } = editUserInfoFormComponentSetup();

        expect(enzymeWrapper.getElements()).toMatchSnapshot();
    });

    it('should render Formik Component', () => {
        const { enzymeWrapper } = editUserInfoFormComponentSetup();

        expect(enzymeWrapper.find(Formik)).toHaveLength(1);
    });

    it('should test if call handleSubmit()', (done) => {
        const { enzymeWrapper, props } = editUserInfoFormComponentSetup();

        enzymeWrapper.find(Formik).simulate('submit');
        
        expect(props.editUserInfoHandler).toHaveBeenCalledTimes(1);
        expect(props.createSnackbar).toHaveBeenCalledTimes(1);

        done();
    });
});