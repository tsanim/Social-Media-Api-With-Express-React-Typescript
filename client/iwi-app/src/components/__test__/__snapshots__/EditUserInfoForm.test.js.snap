// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUserInfoForm Component should render initial layout 1`] = `
Array [
  <Formik
    initialValues={
      Object {
        "form": Object {
          "firstName": "",
          "lastName": "",
          "username": "",
        },
      }
    }
    onSubmit={[Function]}
    validationSchema={
      ObjectSchema {
        "_blacklist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "_conditions": Array [],
        "_defaultDefault": [Function],
        "_deps": Array [],
        "_excludedEdges": Array [],
        "_exclusive": Object {},
        "_mutate": undefined,
        "_nodes": Array [
          "lastName",
          "firstName",
          "username",
        ],
        "_options": Object {
          "abortEarly": true,
          "recursive": true,
        },
        "_type": "object",
        "_typeError": [Function],
        "_whitelist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "fields": Object {
          "firstName": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "min": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "lastName": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "min": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
          "username": StringSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "min": true,
            },
            "_mutate": undefined,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "string",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
      }
    }
  >
    [Function]
  </Formik>,
]
`;
